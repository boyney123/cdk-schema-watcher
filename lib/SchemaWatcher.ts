import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { CfnOutput, RemovalPolicy } from 'aws-cdk-lib';

import { EventBus, EventPattern, Rule } from 'aws-cdk-lib/aws-events';
import { LogGroup } from 'aws-cdk-lib/aws-logs';

export interface SchemaWatcherProps {
  readonly eventPattern?: EventPattern;
  readonly type?: 'Changed' | 'New' | 'All';
  readonly schemas?: string[];
  readonly sources?: string[];
  plugins: any[];
}

const getEventPatternForTypeOfNotification = (type: string, schemas?: string[], sources?: string[]) => {
  let pattern = {};

  if (schemas && schemas.length > 0) {
    pattern = {
      detail: {
        SchemaName: schemas
      },
    };
  }

  if(sources && sources.length > 0){
    pattern = {
      detail: {
        SchemaName: sources.map(source => ({ prefix: source }))
      },
    };
  }

  switch (type) {
    case 'Changed':
      pattern = {
        ...pattern,
        detailType: ['Schema Version Created'],
        source: ['aws.schemas'],
      };
      break;
    case 'New': {
      pattern = {
        ...pattern,
        detailType: ['Schema Created'],
        source: ['aws.schemas'],
      };
      break;
    }
    default:
      pattern = {
        ...pattern,
        detailType: ['Schema Created', 'Schema Version Created'],
        source: ['aws.schemas'],
      };
  }

  return pattern;
};

export class SchemaWatcher extends Construct {
  constructor(scope: Construct, id: string, config: SchemaWatcherProps) {
    super(scope, id);

    const typeOfNotification = config.type || 'All';

    // have to listen for schema changes on the default event bus
    const defaultEventBus = EventBus.fromEventBusName(this, 'EventBus', 'default');

    // useful for debugging
    const logGroup = new LogGroup(this, 'SchemaWatcherLogGroup', {
      logGroupName: `/aws/events/schema-listeners/${id}/schema-notification-logs`,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    // Setup Rule on the bus....
    const rule = new Rule(this, 'Rule', {
      eventBus: defaultEventBus,
      // dynamicly generate event pattern based on consumer inputs
      eventPattern: getEventPatternForTypeOfNotification(typeOfNotification, config.schemas, config.sources),
      ruleName: `${id}-listen-to-schema-changes`,
    });

    rule.addTarget(new cdk.aws_events_targets.CloudWatchLogGroup(logGroup));

    // go through all user defined plugins and add them to the stack, this is for future plugins
    config.plugins.map((Plugin) => {
      new Plugin(this, 'Plugin', {
        rule: rule,
        test: true,
      });
    });

    new CfnOutput(this, 'LogGroup', { value: logGroup.logGroupName });
  }
}
