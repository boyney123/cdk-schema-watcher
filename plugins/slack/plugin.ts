import * as path from 'path';
import { Construct } from 'constructs';

import { Rule } from 'aws-cdk-lib/aws-events';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

import { aws_events_targets, CfnOutput, Duration } from 'aws-cdk-lib';
import { aws_lambda as lambda, Stack } from 'aws-cdk-lib';
import { PolicyStatement, Effect } from 'aws-cdk-lib/aws-iam';

interface PluginConfig {
  readonly rule: Rule;
}

export interface SlackNofiferProps {
  readonly API_KEY?: string;
  readonly CHANNEL_ID?: string;
}

export class SlackNotifier {
  constructor(config: SlackNofiferProps) {
    return class extends Construct {
      constructor(scope: Construct, id: string, pluginConfig: PluginConfig) {
        super(scope, id);

        const region = Stack.of(this).region;
        const account = Stack.of(this).account;

        if(!config.API_KEY || !config.CHANNEL_ID){
          throw new Error('Missing SLACK credentials')
        }

        // Example producer of an event using the enrichment metadata pattern.
        const slackNotificationFunction: NodejsFunction = new NodejsFunction(this, 'slackNotification', {
          memorySize: 1024,
          timeout: Duration.seconds(5),
          runtime: lambda.Runtime.NODEJS_16_X,
          handler: 'handler',
          entry: path.join(__dirname, './process-schema-to-slack-lambda.ts'),
          environment: {
            SLACK_API_KEY: config.API_KEY,
            CHANNEL_ID: config.CHANNEL_ID,
          },
        });

        // allow our slack notification function to get schemas from our event bus
        slackNotificationFunction.addToRolePolicy(
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ['schemas:DescribeSchema'],
            // resources: ['*'],
            resources: [`arn:aws:schemas:${region}:${account}:schema/*`],
          })
        );

        pluginConfig.rule.addTarget(new aws_events_targets.LambdaFunction(slackNotificationFunction));

        new CfnOutput(this, 'slackNotificationARN', { value: slackNotificationFunction.functionArn });
      }
    };
  }
}
