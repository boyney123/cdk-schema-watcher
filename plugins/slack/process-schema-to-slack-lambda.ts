import { ChatPostMessageArguments, WebClient } from '@slack/web-api';
import { SchemasClient, DescribeSchemaCommand } from '@aws-sdk/client-schemas';
import { EventBridgeEvent } from 'aws-lambda'
import diffString from 'variable-diff';

const schemaClient = new SchemasClient({});

const client = new WebClient(process.env.SLACK_API_KEY, {});

const buildURLForSchema = (event: any, version: string) => {
  return `https://${event.region}.console.aws.amazon.com/events/home?region=${event.region}#/registries/${event.detail.RegistryName}/schemas/${event.detail.SchemaName}/version/${version}`;
};

export async function handler(event: EventBridgeEvent<'Schema Version Created' | 'Schema Created', any>) {
  let blocks;

  if (!process.env.CHANNEL_ID) {
    throw Error('No Channel ID found');
  }

  if (event['detail-type'] === 'Schema Version Created') {
    blocks = await buildBlocksForSchemaChanged(event);
  } else {
    blocks = await buildBlocksForNewSchema(event);
  }

  try {
    // write message to slack
    await client.chat.postMessage(blocks as ChatPostMessageArguments);
  } catch (error) {
    console.log(error);
  }
}

const buildBlocksForSchemaChanged = async (event: any) => {
  const [source, detailType] = event.detail.SchemaName.split('@');
  const newVersion = event.detail.Version;
  const oldVersion = newVersion - 1;

  const { Content: oldVersionSchema } = await schemaClient.send(
    new DescribeSchemaCommand({
      RegistryName: event.detail.RegistryName,
      SchemaName: event?.detail.SchemaName,
      SchemaVersion: oldVersion.toString(),
    })
  );

  const { Content: newVersionSchema } = await schemaClient.send(
    new DescribeSchemaCommand({
      RegistryName: event.detail.RegistryName,
      SchemaName: event?.detail.SchemaName,
      SchemaVersion: newVersion,
    })
  );

  if (!oldVersionSchema || !newVersionSchema) {
    throw new Error('Cannot find old and new version of schema');
  }
  

  const { text: diff } = diffString(JSON.parse(oldVersionSchema), JSON.parse(newVersionSchema), {
    color: false,
  });

  return {
    channel: process.env.CHANNEL_ID,
    blocks: [
      {
        type: 'divider',
      },
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text: `üö® New schema version found for event *"${detailType}"*`,
        },
      },
      {
        type: 'section',
        fields: [
          {
            type: 'mrkdwn',
            text: `*Previous Version:*\n<${buildURLForSchema(event, oldVersion.toString())}|version ${oldVersion}>`,
          },
          {
            type: 'mrkdwn',
            text: `*New Version:*\n<${buildURLForSchema(event, newVersion.toString())}|version ${newVersion}>`,
          },
        ],
      },
      {
        type: 'section',
        fields: [
          {
            type: 'mrkdwn',
            text: `*Event Source:*\n${source}`,
          },
          {
            type: 'mrkdwn',
            text: `*Creation Date:*\n${event.detail.CreationDate}`,
          },
        ],
      },
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text: `*Diff (v${oldVersion}:v${newVersion})* \`\`\`${diff}\`\`\``,
        },
      },
      {
        type: 'context',
        elements: [
          {
            type: 'mrkdwn',
            text: '_What am I seeing this? A new version of an event has been created. You may have downstream consumers that may break due to contract changes._',
          },
        ],
      },
      {
        type: 'divider',
      },
    ],
  };
};

const buildBlocksForNewSchema = async (event: any) => {
  const [source, detailType] = event.detail.SchemaName.split('@');
  const newVersion = event.detail.Version;

  const { Content: schema } = await schemaClient.send(
    new DescribeSchemaCommand({
      RegistryName: event.detail.RegistryName,
      SchemaName: event?.detail.SchemaName,
      SchemaVersion: newVersion,
    })
  );

  if (!schema) {
    throw new Error('No schema found for event');
  }

  return {
    channel: process.env.CHANNEL_ID,
    blocks: [
      {
        type: 'divider',
      },
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text: `‚≠êÔ∏è A new event has been published *"${detailType}"*`,
        },
      },
      {
        type: 'section',
        fields: [
          {
            type: 'mrkdwn',
            text: `*Version:*\n<${buildURLForSchema(event, newVersion)}|version ${newVersion}>`,
          },
          {
            type: 'mrkdwn',
            text: `*SchemaType:*\n${event.detail.SchemaType}`,
          },
        ],
      },
      {
        type: 'section',
        fields: [
          {
            type: 'mrkdwn',
            text: `*Event Source:*\n${source}`,
          },
          {
            type: 'mrkdwn',
            text: `*Creation Date:*\n${event.detail.CreationDate}`,
          },
        ],
      },
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text: `*Schema (v${newVersion})* \`\`\`${JSON.stringify(JSON.parse(schema), null, 4)}\`\`\``,
        },
      },
      {
        type: 'context',
        elements: [
          {
            type: 'mrkdwn',
            text: '_What am I seeing this? A new event has been created on your event bus._',
          },
        ],
      },
      {
        type: 'divider',
      },
    ],
  };
};