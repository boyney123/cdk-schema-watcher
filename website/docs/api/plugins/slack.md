---
sidebar_position: 1
id: slack
title: 'ðŸ“¦ Slack Plugin'
---

SchemaWatcher allows you to notify your downstream consumers of schema changes using Slack.

You can pick the channel you want to notify.

## Getting started

First you will need a `SLACK_API_KEY` and `SLACK_CHANNEL_ID` you can find out how to get [these on the slack documentation](https://slack.com/help/articles/215770388-Create-and-regenerate-API-tokens).

Once you have your `SLACK_API_KEY` and `SLACK_CHANNEL_ID` you can configure the Slack plugin.

First import the plugin.

```js
import { SlackNotifier } from 'cdk-schema-watcher/plugins';
```

Next, configure SchemaWatcher to use the plugin.

```js
export class CdkSchemaWatcherSandboxStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new SchemaWatcher(this, 'MyTeam', {
      type: 'All',
      schemas: ['myapp.users@UserCreated'],
      sources: ['myapp.users', 'myapp.orders'],
      plugins: [
        new SlackNotifier({
          API_KEY: process.env.SLACK_API_KEY,
          CHANNEL_ID: process.env.SLACK_CHANNEL_ID,
        }),
      ],
    });

  }
}
```

:::tip
When running `cdk deploy` you will need to set the `SLACK_API_KEY` and `SLACK_CHANNEL_ID` as environmental variables. 

If you would to prefer to use another service to pass these values into the construct you can. You can pass the configuration into the construct anyway you want.
:::

Once you deploy the construct it will start listening to your configured schemas.

## Example output

The Slack plugin will output two different styles of posts `New Schemas` and `Changed Schemas`.

#### Example of `New Schema Version` into slack

When a new version of a schema is raised, a diff is posted back to Slack. The diff between the new and old schemas.

![new schema slack post](/img/slack-example1.png)

#### Example of `New Schema` into slack

When new schemas are raised the Schema is posted to slack with links back to your AWS account.

![new schema slack post](/img/slack-example2.png)