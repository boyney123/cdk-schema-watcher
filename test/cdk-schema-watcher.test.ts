import { SynthUtils, expect as expectCDK, haveResourceLike } from '@aws-cdk/assert';
import { Stack } from 'aws-cdk-lib';
import { SchemaWatcher } from '../lib';
import { SlackNotifier } from '../plugins/slack/plugin';

describe('SchemaWatcher', () => {
  it('snapshot test SchemaWatcher default params', () => {
    const stack = new Stack();
    new SchemaWatcher(stack, 'MyTestTeam', {
      schemas: ['myapp.users@Test'],
      type: 'All',
      plugins: [
        new SlackNotifier({
          API_KEY: 'test',
          CHANNEL_ID: 'test',
        }),
      ],
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  it('cdk-schema-watcher creates a default log group for all configured events', () => {
    const stack = new Stack();

    new SchemaWatcher(stack, 'MyTestTeam', {
      schemas: ['myapp.users@Test'],
      type: 'All',
      plugins: [
        new SlackNotifier({
          API_KEY: 'test',
          CHANNEL_ID: 'test',
        }),
      ],
    });

    expectCDK(stack).to(
      haveResourceLike('AWS::Logs::LogGroup', {
        LogGroupName: '/aws/events/schema-listeners/MyTestTeam/schema-notification-logs',
        RetentionInDays: 731,
      })
    );
  });

  it('cdk-schema-watcher creates a rule onto the default EventBridge event bus to listen for user configured schemas', () => {
    const stack = new Stack();

    new SchemaWatcher(stack, 'MyTestTeam', {
      schemas: ['myapp.users@Test'],
      type: 'All',
      plugins: [
        new SlackNotifier({
          API_KEY: 'test',
          CHANNEL_ID: 'test',
        }),
      ],
    });

    expectCDK(stack).to(
      haveResourceLike('AWS::Events::Rule', {
        EventBusName: 'default',
        EventPattern: {
          detail: {
            SchemaName: ['myapp.users@Test'],
          },
          'detail-type': ['Schema Created', 'Schema Version Created'],
          source: ['aws.schemas'],
        },
        Name: 'MyTestTeam-listen-to-schema-changes',
        State: 'ENABLED',
      })
    );
  });
  
});
