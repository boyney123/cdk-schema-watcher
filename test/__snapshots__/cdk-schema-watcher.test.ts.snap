// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaWatcher snapshot test SchemaWatcher default params 1`] = `
Object {
  "Outputs": Object {
    "MyTestTeamLogGroup5C7602E6": Object {
      "Value": Object {
        "Ref": "MyTestTeamSchemaWatcherLogGroup345F640D",
      },
    },
    "MyTestTeamPluginslackNotificationARN7A406A0A": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "MyTestTeamPluginslackNotificationD0D29E5F",
          "Arn",
        ],
      },
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e845402ce43b66fc6f20df4a239f20f8662eb6c7f920b94cf6542dd0e64ce0f7.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EventsLogGroupPolicyMyTestTeamRule0272B20046BD0CD0": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "EventsLogGroupPolicyMyTestTeamRule0272B200CustomResourcePolicyFE182C61",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"CloudWatchLogs\\",\\"action\\":\\"putResourcePolicy\\",\\"parameters\\":{\\"policyName\\":\\"EventsLogGroupPolicyMyTestTeamRule0272B200\\",\\"policyDocument\\":\\"{\\\\\\"Statement\\\\\\":[{\\\\\\"Action\\\\\\":[\\\\\\"logs:PutLogEvents\\\\\\",\\\\\\"logs:CreateLogStream\\\\\\"],\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\",\\\\\\"Principal\\\\\\":{\\\\\\"Service\\\\\\":\\\\\\"events.amazonaws.com\\\\\\"},\\\\\\"Resource\\\\\\":\\\\\\"",
              Object {
                "Fn::GetAtt": Array [
                  "MyTestTeamSchemaWatcherLogGroup345F640D",
                  "Arn",
                ],
              },
              "\\\\\\"}],\\\\\\"Version\\\\\\":\\\\\\"2012-10-17\\\\\\"}\\"},\\"physicalResourceId\\":{\\"id\\":\\"EventsLogGroupPolicyMyTestTeamRule0272B200\\"}}",
            ],
          ],
        },
        "Delete": "{\\"service\\":\\"CloudWatchLogs\\",\\"action\\":\\"deleteResourcePolicy\\",\\"parameters\\":{\\"policyName\\":\\"EventsLogGroupPolicyMyTestTeamRule0272B200\\"},\\"ignoreErrorCodesMatching\\":\\"400\\"}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"CloudWatchLogs\\",\\"action\\":\\"putResourcePolicy\\",\\"parameters\\":{\\"policyName\\":\\"EventsLogGroupPolicyMyTestTeamRule0272B200\\",\\"policyDocument\\":\\"{\\\\\\"Statement\\\\\\":[{\\\\\\"Action\\\\\\":[\\\\\\"logs:PutLogEvents\\\\\\",\\\\\\"logs:CreateLogStream\\\\\\"],\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\",\\\\\\"Principal\\\\\\":{\\\\\\"Service\\\\\\":\\\\\\"events.amazonaws.com\\\\\\"},\\\\\\"Resource\\\\\\":\\\\\\"",
              Object {
                "Fn::GetAtt": Array [
                  "MyTestTeamSchemaWatcherLogGroup345F640D",
                  "Arn",
                ],
              },
              "\\\\\\"}],\\\\\\"Version\\\\\\":\\\\\\"2012-10-17\\\\\\"}\\"},\\"physicalResourceId\\":{\\"id\\":\\"EventsLogGroupPolicyMyTestTeamRule0272B200\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::CloudwatchLogResourcePolicy",
      "UpdateReplacePolicy": "Delete",
    },
    "EventsLogGroupPolicyMyTestTeamRule0272B200CustomResourcePolicyFE182C61": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "logs:PutResourcePolicy",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "logs:DeleteResourcePolicy",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventsLogGroupPolicyMyTestTeamRule0272B200CustomResourcePolicyFE182C61",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyTestTeamPluginslackNotificationD0D29E5F": Object {
      "DependsOn": Array [
        "MyTestTeamPluginslackNotificationServiceRoleDefaultPolicy0BF0B9DA",
        "MyTestTeamPluginslackNotificationServiceRoleADC4B501",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "944806ecbdf260545f83045489c35f66de16e48ffcef57ba6b39daf2da654e53.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CHANNEL_ID": "test",
            "SLACK_API_KEY": "test",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestTeamPluginslackNotificationServiceRoleADC4B501",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestTeamPluginslackNotificationServiceRoleADC4B501": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestTeamPluginslackNotificationServiceRoleDefaultPolicy0BF0B9DA": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "schemas:DescribeSchema",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:schemas:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":schema/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyTestTeamPluginslackNotificationServiceRoleDefaultPolicy0BF0B9DA",
        "Roles": Array [
          Object {
            "Ref": "MyTestTeamPluginslackNotificationServiceRoleADC4B501",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyTestTeamRule6983D7A1": Object {
      "Properties": Object {
        "EventBusName": "default",
        "EventPattern": Object {
          "detail": Object {
            "SchemaName": Array [
              "myapp.users@Test",
            ],
          },
          "detail-type": Array [
            "Schema Created",
            "Schema Version Created",
          ],
          "source": Array [
            "aws.schemas",
          ],
        },
        "Name": "MyTestTeam-listen-to-schema-changes",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":logs:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":log-group:",
                  Object {
                    "Ref": "MyTestTeamSchemaWatcherLogGroup345F640D",
                  },
                ],
              ],
            },
            "Id": "Target0",
          },
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "MyTestTeamPluginslackNotificationD0D29E5F",
                "Arn",
              ],
            },
            "Id": "Target1",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "MyTestTeamRuleAllowEventRuleMyTestTeamPluginslackNotificationCB65424282522DCF": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "MyTestTeamPluginslackNotificationD0D29E5F",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "MyTestTeamRule6983D7A1",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyTestTeamSchemaWatcherLogGroup345F640D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": "/aws/events/schema-listeners/MyTestTeam/schema-notification-logs",
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
